server/src/infra/repositories/database.repository.ts

```diff
@@ -25,4 +28,34 @@ export class DatabaseRepository implements IDatabaseRepository {
   async runMigrations(options?: { transaction?: 'all' | 'none' | 'each' }): Promise<void> {
     await this.dataSource.runMigrations(options);
   }
+
+  async withLock<R>(lock: DatabaseLock, callback: () => Promise<R>): Promise<R> {
+    let res;
+    await this.asyncLock.acquire(DatabaseLock[lock], async () => {
```

Comment:
Why can't this just be "lock"?

Reply:
There are two kinds of locks being used here, so I thought it'd be good to specify the kind in the name. Not really needed though since it's already an `AsyncLock`.

Reply:
I meant you are unexpected into an enum with an enum

Reply:
Oh, that was a typing thing: `DatabaseLock is not assignable to parameter of type 'string | string[]`

Reply:
Hmm. We probably want to just cast it in that case.